### ISSUE TITLE
Calling TrackPlayer.Destroy() in useEffect cleanup function crashes ios and "fries" android on 2.2.0-rc4

### ISSUE THREAD

**Issue by marcfrankel** [doublesymmetry/react-native-track-player#1587]
**Describe the Bug**

On the page where we use the track player we have a useEffect cleanup function that when the page is unloaded the TrackPlayer is destroyed. On RC4, calling this function on iOS causes an immediate hard crash. The type that triggers a popup from MacOs saying the app quit unexpectedly. On Android it technically doesn't crash, but the app begins to display extremely weird behavior afterwards. For instance, despite not saying so, it loses connection to metro. At least so far as fast refresh goes. Additionally, we are using Firebase and all of our Firestore listeners stop working. Like I said... weird behavior.

**Steps To Reproduce**
How can someone else reproduce this bug?

Call TrackPlayer.Destroy() during unmounting of the track player on RC4

**Code To Reproduce**
Please provide a simple code example that allows others to replicate the bug.

`
useEffect(() => {
    return () => {  
            TrackPlayer.destroy();
        }; 
}, []);  
`

**Environment Info:**
Paste the results of `npx react-native info`
System:
    OS: macOS 12.4
    CPU: (12) x64 Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
    Memory: 791.21 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.2 - /usr/local/bin/node
    Yarn: 1.22.19 - ~/.yarn/bin/yarn
    npm: 8.1.2 - /usr/local/bin/npm
    Watchman: 2022.03.14.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.2 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    Android SDK:
      API Levels: 29, 30, 31
      Build Tools: 29.0.2, 30.0.2, 31.0.0, 32.0.0
      System Images: android-29 | Intel x86 Atom, android-29 | Intel x86 Atom_64, android-29 | Google APIs Intel x86 Atom, android-29 | Google APIs Intel x86 Atom_64
      Android NDK: Not Found
  IDEs:
    Android Studio: 2021.2 AI-212.5712.43.2112.8512546
    Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.13 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.68.2 => 0.68.2 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
info React Native v0.69.1 is now available (your project is running on v0.68.2).

Paste the exact `react-native-track-player` version you are using
Exact same code works on 2.2.0-RC3, the crash occurs on 2.2.0-RC4

Real device? Or simulator?
Both (iOS and Android)

What OS are you running?
MacOS

**How I can Help**
What can you do to help resolve this?
Have you investigated the underlying JS or Swift/Android code causing this bug?
Can you create a Pull Request with a fix?

There's a potential this is related to [this](https://github.com/doublesymmetry/react-native-track-player/issues/1585) issue, but for now, I suggest to anyone encountering this issue just roll back the version.

I can post the crash log from MacOs, but to be honest, I spent hours trying to figure it out before posting here and I didn't see anything worth while in it. But happy to share it if you think it can help!

**Comment by marcfrankel** [comment]
Update: Rolling back to RC3 only fixes iOS, Android remains "fried"

**Comment by jspizziri** [comment]
@marcfrankel I'm not able to reproduce your issue. When I call `destroy` I do not get any fatal exception. I tested on both Android & iOS.

Please create a reproducible example by forking the project and updating the code in `example` to demonstrate your issue.

**Comment by DennisdeWitNL** [comment]
The .reset(); crash on iOS is occurrd by an ‘index out of range’ error in handleAudioPlayerQueueIndexChange. It happens when there is only one track in the queue (namely the current track)

This is the code that seems to cause the error

`if let nextIndex = nextIndex, nextIndex < player.items.count {
            let track = player.items[nextIndex]
            isTrackLiveStream = (track as? Track)?.isLiveStream ?? false
        }`

I have commented out that code and the queue succesfully resets without crashing. It seems the code assumes there is a next index, even when there is not. Thus it crashes because the index is out of range. 

Haven’t tested Android yet, as I want some sleep now. :)

**Comment by jspizziri** [comment]
@DennisdeWitNL

This is a separate issue. The reset issue has already been solved on iOS via another ticket (please see the PR I linked above).

**Comment by DennisdeWitNL** [comment]
@jspizziri Look at the code above. It seems the code is the same as your PR, but it still didn’t fix the issue. Or am I really so tired that I am misreading your PR?

**Comment by jspizziri** [comment]
@DennisdeWitNL this is the wrong ticket to discuss reset, as it's about destroy.

Please hop onto this one https://github.com/doublesymmetry/react-native-track-player/issues/1585

**Comment by cashtotalszero** [comment]
+1 on this. Both the `reset() `and the `destroy()` are causing iOS crashes while on the following setup:

```
 "react": "17.0.2",
 "react-native": "0.68.2", 
 "react-native-track-player": "2.2.0-rc4",
```

Code snippet from my component:

```       
 useEffect(() => {
   TrackPlayer.addEventListener(Event.PlaybackError, () =>
      setHasError(true),
    )
    
    return () => {
      TrackPlayer.destroy() // Causes a crash

       // TrackPlayer.reset() // Also causes a crash :(
     }
 }, [])
```

From the discussions above it looks like you have found the problem though. Is there a patch release incoming?

**Comment by jspizziri** [comment]
The reset issue has been fixed on another ticket. The destroy issue I have not been able to reproduce. Please try to create a repro based on the example project.

I don't have the ability to release. But I hope for one in the next week or two. I suggest using RC3 if this is a showstopper for you.

**Comment by cashtotalszero** [comment]
I have managed to repro using the example code (which I have embedded into a screen within my existing app). All example code was copied and pasted directly and plays the tracks as expected.

To see the crash - the only change is adding a cleanup function to the effect in `App.tsx`:

```
    useEffect(() => {
        async function run() {
            const isSetup = await SetupService()
            setIsPlayerReady(isSetup)

            const queue = await TrackPlayer.getQueue()
            if (isSetup && queue.length <= 0) {
                await QueueInitalTracksService()
            }
        }

        run()

        return () => {
            TrackPlayer.destroy()
        }
    }, [])
```
This is causing iOS to crash every time I navigate away from the screen showing the audio player.

**Comment by jspizziri** [comment]
@cashtotalszero

I need you to send me a link to a forked of this repo with the reproduction added as a commit. Thanks!

**Comment by ildfreelancer** [comment]
@jspizziri sorry to ping you if it's night.

Take a look at @cashtotalszero's code above, this issue comes from with SetupService
in the example,
https://github.com/doublesymmetry/react-native-track-player/blob/d961abe733a77ba9d5b0cffb193cb864eaf0c784/example/src/services/SetupService.ts

after TrackPlayer.destroy is called, TrackPlayer.getCurrentTrack() will cause the crash due to  player is destroyed.

**Comment by jspizziri** [comment]
As I've mentioned a few times now. I cannot reproduce this despite my best efforts. I'm going to close this, if someone is having the issue still I'll happily reopen it if they provide a fork of the repository with a reproduction in the example project.

**Comment by pankaj9singh** [comment]
Still same issue with iOS, crashing on destroy and reset

**Comment by pankaj9singh** [comment]
It worked thank you 

https://github.com/doublesymmetry/react-native-track-player/issues/1585#issuecomment-1194163796

**Comment by Harukisatoh** [comment]
To summarize for future readers, this bug happens on versions 2.2.0-rc3 and 2.2.0-rc4, and it seems related to the issue #1585. This other bug was already solved and by consequence this one is too.

The solution isn't released yet, so if you want to solve this on your end you have two options: 
1. Downgrade to versions below 2.2.0-rc3 (I don't know in which one the bug was introduced, so I'm not sure which one you have to downgrade to)
2. Or update the package to the nightly version ([click here for more info](https://react-native-track-player.js.org/docs/next/basics/installation#unstable--nightly))

By the way, I guess that @jspizziri wasn't able to reproduce the issue because it was testing in the nightly version, and in that version, the issue was already solved (but anyway, the bug is solved 🎉)

**Comment by devpolo** [comment]
Looks like it's fixed on the latest `2.2.0-rc4-e36a016d1640e75649d6da4056f8977da33864e1` version. 

You can upgrade it with :
`yarn upgrade react-native-track-player@2.2.0-rc4-e36a016d1640e75649d6da4056f8977da33864e1`
